plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs

```diff
@@ -49,29 +51,39 @@ data State s = State { stateData :: s, stateValue :: Value }
 -- of the transition in the context of the current transaction.
 data StateMachine s i = StateMachine {
       -- | The transition function of the state machine. 'Nothing' indicates an invalid transition from the current state.
-      smTransition :: State s -> i -> Maybe (TxConstraints Void Void, State s),
+      smTransition  :: State s -> i -> Maybe (TxConstraints Void Void, State s),
 
       -- | Check whether a state is the final state
-      smFinal      :: s -> Bool,
+      smFinal       :: s -> Bool,
 
       -- | The condition checking function. Can be used to perform
       --   checks on the pending transaction that aren't covered by the
       --   constraints. 'smCheck' is always run in addition to checking the
       --   constraints, so the default implementation always returns true.
-      smCheck      :: s -> i -> ValidatorCtx -> Bool
+      smCheck       :: s -> i -> ValidatorCtx -> Bool,
+
+      -- | The 'Currency' of the thread token that identifies the contract instance.
+      smThreadToken :: Maybe Currency
```

Comment:
Should this really be part of the *instance*, given that part of the point is to identify different state machines?

Reply:
IMO it doesn't matter if it's in the instance or in the datum. You need to pass around the same information in both cases. But I can move it to the datum.

Reply:
I meant the `StateMachineInstance` type!

Reply:
I was thinking something like: what do I need to produce an *instance* of a state machine? I need to give it:
- The state machine definition
- The validator script for it
- The unique output I'm going to spend to start it (which implies the thread token currency)

Reply:
> I meant the StateMachineInstance type!

I see. The term is overloaded. `StateMachineInstance` only exists offchain, but we also need to keep track of the token somewhere in the on-chain code, don't we?

> The unique output I'm going to spend to start it (which implies the thread token currency)

In this version I deliberately made the assumption that the state token would be passed in from the outside. (See PR description). But maybe the change to having it generate and destroy the token automatically isn't so big. I'll have a look.